[doc] Update syntax.rst and related sections (#967)
taichi-dev/taichi: docs/syntax.rst

    res = add_xy(2.3, 1.1)
    print(res)  # 3, since return type is ti.i32

.. note::

>     For now, we only support one scalar as return value. Returning ``ti.Matrix`` or ``ti.Vector`` is not supported. Python-style tuple return is not supported either. For example,

Reviewer: ```suggestion
    For now, we only support one scalar as return value. Returning ``ti.Matrix`` or ``ti.Vector`` is not supported. Python-style tuple return is not supported either. e.g.:
```

Author: Why is `e.g.:` better? In English people usually use `E.g.,`, which is the same as `For example,`.

Reviewer: ```suggestion
    For now, we only support one scalar as return value. Returning ``ti.Matrix`` or ``ti.Vector`` is not supported. Python-style tuple return is not supported either. For example:
```
Fine, but I found that `poedit` hate this, when en ends with `,`, and cn ends with `：`, it WARNs you incorrect, and you can't get rid that WARN... Let's use `:` for `e.g.`'s, which helps both en and cn, wdyt?

Author: Thanks. Sounds good to me. No perfect solution here I guess.

    .. code-block:: python

        @ti.kernel
        def bad_kernel() -> ti.Matrix:
            return ti.Matrix([[1, 0], [0, 1]])  # Error

        @ti.kernel
        def bad_kernel() -> (ti.i32, ti.f32):
            x = 1
            y = 0.5
            return x, y  # Error

> We also support **template arguments** (see :ref:`template_metaprogramming`) and **external array arguments** (see :ref:`external`) in Taichi kernels.

Reviewer: Let's move this line to L15 (below the `kernel can have arguments, arguments type-hinted...`).

Author: I would actually stick to this organization because moving this ahead might cause a digression for the reader, who are simply learning basic kernel usages. WDYT?

.. warning::

   When using differentiable programming, there are a few more constraints on kernel structures. See the **Kernel Simplicity Rule** in :ref:`differentiable`.

   Also, please do not use kernel return values in differentiable programming, since the return value will not be tracked by automatic differentiation. Instead, store the result into a global variable (e.g. ``loss[None]``).

Functions
---------

> Use ``@ti.func`` to decorate your Taichi functions. These functions are callable only in `Taichi`-scope. Do not call them in `Python`-scopes.

Reviewer: ```suggestion
Use ``@ti.func`` to decorate your Taichi functions. These functions are callable only in `Taichi`-scope. Do not call them in `Python`-scope.
```
Let's **always keep singular form**, unless it's clarified by context. I guess that's just common for English to lack a netural version for a word, like `man` and `woman`, and you have to choose one of them to stand for both...

Author: I feel like we should always use `scopes`. The singular form must be `a scope` or `the scope` in a sentence. We should use plural forms consistently since we are not talking about a specific scope.

Author: Actually, either option is fine as long as we use it consistently across the documentation.

Reviewer: ```suggestion
Use ``@ti.func`` to decorate your Taichi functions. These functions are callable only in `Taichi`-scopes. Do not call them in `Python`-scopes.
```

Author: How about this: given there are many places in the document where `Taichi-scope` is already used, let's use the singular form as you suggested throughout the documentation. In a separate PR I'll do a pass to unify all usage of `scope`.

Reviewer: Ack :)
